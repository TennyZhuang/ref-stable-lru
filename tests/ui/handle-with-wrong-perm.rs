use ref_stable_lru::LruCache;
use std::num::NonZeroUsize;

fn main() {
    let mut cache1: LruCache<&'static str, String> = LruCache::new(NonZeroUsize::new(3).unwrap());

    cache1.scope(|mut handle1, perm1| {
        let mut cache2: LruCache<&'static str, String> =
            LruCache::new(NonZeroUsize::new(3).unwrap());

        // We don't really want to use `handle2`, but construct a fake `ValuePerm` and try to modify `cache1`.
        cache2.scope(|_handle2, mut perm2| {
            let x = handle1.get(&"a", &perm1).unwrap().as_str();
            let y = handle1.get(&"b", &perm1).unwrap().as_str();
            let z = handle1.get(&"c", &perm1).unwrap().as_str();
            // Should fail here due to lifetime (`'brand`) not match.
            handle1.put("a", "".to_string(), &mut perm2);
            [x, y, z].join(" ")
        })
    });
}
