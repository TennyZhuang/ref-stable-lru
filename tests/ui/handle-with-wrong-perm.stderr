error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/handle-with-wrong-perm.rs:17:13
   |
7  |     cache1.scope(|mut handle1, perm1| {
   |                   ----------- `handle1` declared here, outside of the closure body
...
12 |         cache2.scope(|_handle2, mut perm2| {
   |                       -------- `_handle2` is a reference that is only valid in the closure body
...
17 |             handle1.put("a", "".to_string(), &mut perm2);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `_handle2` escapes the closure body here
   |
   = note: requirement occurs because of a mutable reference to `ValuePerm<'_>`
   = note: mutable references are invariant over their type parameter
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error[E0521]: borrowed data escapes outside of closure
  --> tests/ui/handle-with-wrong-perm.rs:17:13
   |
7  |     cache1.scope(|mut handle1, perm1| {
   |                   -----------
   |                   |
   |                   `handle1` is a reference that is only valid in the closure body
   |                   has type `CacheHandle<'_, '1, &str, String>`
...
17 |             handle1.put("a", "".to_string(), &mut perm2);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |             |
   |             `handle1` escapes the closure body here
   |             argument requires that `'1` must outlive `'static`
